src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs

```diff
@@ -443,10 +484,51 @@ internal static bool IsNonNullable(IEnumerable<object> attributes)
                 flags.Length >= 0 &&
                 flags[0] == 1) // First element is the property/parameter type.
             {
-                return true;
+                isNullable = true;
+                return true; // [Nullable] found and type is an NNRT
             }
 
-            return false;
+            isNullable = false;
+            return true; // [Nullable] found but type is not an NNRT
+        }
+
+        internal static bool IsNullableBasedOnContext(Type containingType, MemberInfo member)
+        {
+            var attributes = member?.GetCustomAttributes(inherit: true) ?? Array.Empty<object>();
+            var isNullable = AttributesHasNullableContext(attributes);
+            if (isNullable != null)
+            {
+                return isNullable.Value;
+            }
+
+            attributes = containingType.GetCustomAttributes(inherit: false);
```

Comment:
The `inherits: false` warrants a commit. We generally don't do this elsewhere.

Reply:
sure. Will add comments for these - `[Nullable]` and `[NullableContext]` aren't inherited, so there's no value in doing the extra work.